storage/series/reader.go

```diff
@@ -0,0 +1,250 @@
+
+	if end.Before(start) {
+		return nil, xerrors.NewInvalidParamsError(errSeriesReadInvalidRange)
+	}
+
+	var (
+		nowFn        = r.opts.ClockOptions().NowFn()
+		now          = nowFn()
+		cachePolicy  = r.opts.CachePolicy()
+		ropts        = r.opts.RetentionOptions()
+		size         = ropts.BlockSize()
+		alignedStart = start.Truncate(size)
+		alignedEnd   = end.Truncate(size)
+	)
+	if alignedEnd.Equal(end) {
+		// Move back to make range [start, end)
+		alignedEnd = alignedEnd.Add(-1 * size)
+	}
+	// Squeeze the lookup window by what's available to make range queries like [0, infinity) possible
+	earliest := retention.FlushTimeStart(ropts, now)
+	if alignedStart.Before(earliest) {
+		alignedStart = earliest
+	}
+	latest := now.Add(ropts.BufferFuture()).Truncate(size)
+	if alignedEnd.After(latest) {
+		alignedEnd = latest
+	}
+
+	var (
+		first, last = alignedStart, alignedEnd
+		clonedID    ts.ID
+	)
+	for blockAt := first; !blockAt.After(last); blockAt = blockAt.Add(size) {
+		if seriesBlocks != nil {
+			if block, ok := seriesBlocks.BlockAt(blockAt); ok {
+				// Block served from in-memory or in-memory metadata
+				// will defer to disk read
+				stream, err := block.Stream(ctx)
+				if err != nil {
+					return nil, err
+				}
+				if stream != nil {
+					results = append(results, []xio.SegmentReader{stream})
+					// NB(r): Mark this block as read now
+					block.SetLastReadTime(now)
+				}
+				continue
+			}
+		}
+
+		switch {
+		case cachePolicy == CacheAll:
+			// No-op, block metadata should have been in-memory
+		case cachePolicy == CacheAllMetadata:
+			// No-op, block metadata should have been in-memory
+		case r.retriever != nil:
+			// Try to stream from disk
+			if r.retriever.IsBlockRetrievable(blockAt) {
+				if clonedID == nil {
+					// Clone ID as the block retriever uses the ID async so we cannot
+					// be sure about owner not finalizing the cloneableID passed to
+					// the Reader
+					clonedID = r.opts.IdentifierPool().Clone(r.cloneableID)
+					ctx.RegisterFinalizer(clonedID)
+				}
+				// TODO(r): Specify a non-nil OnRetrieveBlock callback to cache the
+				// block back to a series in the shard on a read.
+				var onRetrieve block.OnRetrieveBlock
+				stream, err := r.retriever.Stream(clonedID, blockAt, onRetrieve)
+				if err != nil {
+					return nil, err
+				}
+				if stream != nil {
+					results = append(results, []xio.SegmentReader{stream})
+				}
+			}
+		}
+	}
+
+	if seriesBuffer != nil {
+		bufferResults := seriesBuffer.ReadEncoded(ctx, start, end)
+		if len(bufferResults) > 0 {
+			results = append(results, bufferResults...)
+		}
+	}
+
+	return results, nil
+}
+
+// FetchBlocks returns data blocks given a list of block start times using
+// just a block retriever.
+func (r Reader) FetchBlocks(
+	ctx context.Context,
+	starts []time.Time,
+) ([]block.FetchBlockResult, error) {
+	return r.fetchBlocksWithBlocksMapAndBuffer(ctx, starts, nil, nil)
+}
+
+func (r Reader) fetchBlocksWithBlocksMapAndBuffer(
```

Comment:
Similar to my comment above, a little confused as to what this function is for and why these access pattern is different than what we do to serve reads

Reply:
As mentioned above, this was because we don't want to create a Series when we want to service a read for a series that isn't in memory. It's moving the stitching of results together from the Series to an independent party that can be used directly in the shard (without needing to create a Series when one doesn't exist in memory).